packages/@lwc/engine-dom/src/styles.ts

```diff
@@ -0,0 +1,146 @@
+// Detect IE, via https://stackoverflow.com/a/9851769
+const isIE11 = !isUndefined((document as any).documentMode);
+
+//
+// Style sheet cache
+//
+
+// Global cache of style elements used for fast cloning
+let styleElements: Map<string, HTMLStyleElement> = new Map();
+// Global cache of CSSStyleSheets because these need to be unique based on content
+let stylesheets: Map<string, CSSStyleSheet> = new Map();
+// Bookkeeping of targets to CSS that has already been injected into them, so we don't duplicate
+let shadowRootsToInsertedStylesheets: WeakMap<ShadowRoot, Set<string>> = new WeakMap();
+// Same as above, but for the global document to avoid an extra WeakMap lookup for this common case
+let globalInsertedStylesheets: Set<string> = new Set();
+
+//
+// Test utilities
+//
+
+if (process.env.NODE_ENV === 'development') {
+    // @ts-ignore
+    window.__lwcResetGlobalStylesheets = () => {
+        styleElements = new Map();
+        stylesheets = new Map();
+        shadowRootsToInsertedStylesheets = new WeakMap();
+        globalInsertedStylesheets = new Set();
+    };
+}
+
+function isDocument(target: ShadowRoot | Document): target is Document {
+    return !isUndefined((target as Document).head);
+}
+
+function createFreshStyleElement(content: string) {
+    const elm = document.createElement('style');
+    elm.type = 'text/css';
+    elm.textContent = content;
+    return elm;
+}
+
+function createStyleElement(content: string) {
+    // For a mysterious reason, IE11 doesn't like the way we clone <style> nodes
+    // and will render the incorrect styles if we do things that way. It's just
+    // a perf optimization, so we can skip it for IE11.
+    if (isIE11) {
+        return createFreshStyleElement(content);
+    }
+
+    let elm = styleElements.get(content);
+    if (isUndefined(elm)) {
+        // We don't clone every time, because that would be a perf tax on the first time
+        elm = createFreshStyleElement(content);
+        styleElements.set(content, elm);
+    } else {
+        // This `<style>` may be repeated multiple times in the DOM, so cache it. It's a bit
+        // faster to call `cloneNode()` on an existing node than to recreate it every time.
+        elm = elm.cloneNode(true) as HTMLStyleElement;
+    }
+    return elm;
+}
+
+function createOrGetConstructableStylesheet(content: string) {
+    // It's important for CSSStyleSheets to be unique based on their content, so
+    // that adoptedStyleSheets.indexOf(sheet) works
+    let stylesheet = stylesheets.get(content);
+    if (isUndefined(stylesheet)) {
+        stylesheet = new CSSStyleSheet();
+        stylesheet.replaceSync(content);
+        stylesheets.set(content, stylesheet);
+    }
+    return stylesheet;
+}
+
+function insertConstructableStylesheet(content: string, target: ShadowRoot | Document) {
+    const stylesheet = createOrGetConstructableStylesheet(content);
+    const { adoptedStyleSheets } = target;
+    if (supportsMutableAdoptedStyleSheets) {
```

Comment:
I was wondering if it makes sense to have 2 code paths for inserting a constructible stylesheet. I went back to the original PR https://github.com/salesforce/lwc/pull/2683 and realized that the motivation was performance. I think it would make sense to add a note stating this is a performance optimization.

Reply:
Yep the original motivation was actually https://github.com/salesforce/lwc/pull/2460. Using the mutable array (#2683) was also a perf improvement.

Reply:
Added a comment.